### PR TITLE
Properly handle nil Data from HEAD requests

### PR DESCRIPTION
**Pull request by jshier**
This PR adds support for treating `HEAD` requests as successful when they have `nil` response data. It also refactors our existing treatment of `nil` data response codes and adds tests for this new behavior.

### CODE REVIEW

File: Source/AFError.swift [hunk 1]
```diff
@@ -109,15 +109,15 @@ public enum AFError: Error {
...
+    /// - inputDataZeroLength:             The server response data was zero length.
     /// - inputFileNil:                    The file containing the server response did not exist.
     /// - inputFileReadFailed:             The file containing the server response could not be read.
     /// - stringSerializationFailed:       String serialization failed using the provided `String.Encoding`.
     /// - jsonSerializationFailed:         JSON serialization failed with an underlying system error.
     /// - propertyListSerializationFailed: Property list serialization failed with an underlying system error.
     public enum ResponseSerializationFailureReason {
         case inputDataNil
-        case inputDataNilOrZeroLength
+        case inputDataZeroLength
```
Comment (cnoon) [inline Source/AFError.swift]: This isn't a backwards compatible change @jshier...we can't make this type of change until AF5.

File: Source/ResponseSerialization.swift [hunk 2]
```diff
@@ -253,18 +253,19 @@ extension DownloadRequest {
...
     /// Returns a result data type that contains the response data as-is.
     ///
+    /// - parameter request:  The request performed.
     /// - parameter response: The response from the server.
     /// - parameter data:     The data returned from the server.
     /// - parameter error:    The error already encountered if it exists.
     ///
     /// - returns: The result data type.
-    public static func serializeResponseData(response: HTTPURLResponse?, data: Data?, error: Error?) -> Result<Data> {
+    public static func serializeResponseData(request: URLRequest?, response: HTTPURLResponse?, data: Data?, error: Error?) -> Result<Data> {
```
Comment (cnoon) [inline Source/ResponseSerialization.swift]: This isn't a backwards compatible change @jshier. We'll have to find another way to do this that doesn't change a method that can be subclassed.

File: Source/ResponseSerialization.swift [hunk 3]
```diff
@@ -712,5 +727,32 @@ extension DownloadRequest {
...
 
-/// A set of HTTP response status code that do not contain response data.
-private let emptyDataStatusCodes: Set<Int> = [204, 205]
+// MARK: - Empty Data Handling
+
+extension Request {
+    /// A set of HTTP response status code that do not contain response data.
+    fileprivate static let emptyDataStatusCodes: Set<Int> = [204, 205]
+
+
```
Comment (cnoon) [inline Source/ResponseSerialization.swift]: Extra newline.

File: Source/ResponseSerialization.swift [hunk 4]
```diff
@@ -712,5 +727,32 @@ extension DownloadRequest {
...
+    /// A set of `HTTPMethod`s that are not required to contained response data.
+    fileprivate static let emptyDataHTTPMethods: Set<HTTPMethod> = [.head]
+
+
```
Comment (cnoon) [inline Source/ResponseSerialization.swift]: Extra newline.

File: Tests/TLSEvaluationTests.swift [hunk 5]
```diff
@@ -414,7 +422,15 @@ class TLSEvaluationExpiredLeafCertificateTestCase: BaseTestCase {
...
+    #if os(iOS) || os(macOS)
+        if #available(iOS 10.1, macOS 10.12.0, *) {
+            XCTAssertNotNil(error, "error should not be nil")
+        } else {
+            XCTAssertNil(error, "error should be nil")
+        }
+    #else
         XCTAssertNil(error, "error should be nil")
+    #endif
     }
```
Comment (cnoon) [inline Tests/TLSEvaluationTests.swift]: I don't really understand the diffs here. This is all already in the `master` branch. The only thing that isn't is that you've changed `iOS 10.0` to `10.1` from what I can tell. Do you have any insight here?

Comment (jshier) [inline Tests/TLSEvaluationTests.swift]: I branched from that commit, but for some reason my branch now includes it. Removing the commit should get rid of the spurious changes.

File: Source/ResponseSerialization.swift (hunk 2 above)
Comment (jshier) [inline Source/ResponseSerialization.swift]: Unfortunately I don't see any other way to check the request type without passing the `URLRequest` into the serializer.

File: Source/AFError.swift (hunk 1 above)
Comment (jshier) [inline Source/AFError.swift]: Well, I thought I had mentioned it. But I can revert this part of the changes, it just means that reason is a bit misleading.

File: Tests/TLSEvaluationTests.swift (hunk 5 above)
Comment (jshier) [inline Tests/TLSEvaluationTests.swift]: I also fixed the `macOS` check, as it was just `(macOS)` before, which doesn't work but for some reason compiles.

Comment (cnoon) [inline Tests/TLSEvaluationTests.swift]: Cool.

File: Source/AFError.swift (hunk 1 above)
Comment (cnoon) [inline Source/AFError.swift]: That's my bad if you did. I'd say let's punt this to post 4.1.0. Maybe we can figure out a backwards compatible way to handle this feature before AF5.

**Comment by cnoon** [comment]
Hey @jshier, I had a chance to review this more in-depth and I'm afraid we'll have to wait until AF5 to make these changes. Are there any changes in here that are backwards compatible that should still make it into AF 4.x? I remember you saying you found a couple small issues that you thought were incorrect. Maybe with the tests?

I'm going to close this PR out for now. We can revisit in AF 5. I've added a card to our Trello board.

**Pull request closed without merging**